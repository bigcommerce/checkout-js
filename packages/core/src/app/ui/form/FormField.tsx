import { FieldProps } from 'formik';
import { kebabCase } from 'lodash';
import React, { FunctionComponent, memo, ReactNode, useCallback } from 'react';

import BasicFormField from './BasicFormField';
import FormFieldError from './FormFieldError';
import Label from './Label';

export interface FormFieldProps {
    additionalClassName?: string;
    name: string;
    label?: ReactNode | ((fieldName: string) => ReactNode);
    labelContent?: ReactNode;
    footer?: ReactNode;
    id?: string;
    useFloatingLabel?: boolean;
    input(field: FieldProps<string>): ReactNode;
    onChange?(value: string): void;
}

const FormField: FunctionComponent<FormFieldProps> = ({
    additionalClassName,
    labelContent,
    label,
    onChange,
    footer,
    input,
    name,
    id,
    useFloatingLabel,
}) => {
    const renderField = useCallback(
        (props) => (
            <>
                {useFloatingLabel && input(props)}

                {label && (typeof label === 'function' ? label(name) : label)}
                {labelContent && !label && (
                    <Label
                        htmlFor={name}
                        id={`${id ?? name}-label`}
                        useFloatingLabel={useFloatingLabel}
                    >
                        {labelContent}
                    </Label>
                )}

                {!useFloatingLabel && input(props)}

                <FormFieldError
                    errorId={`${id ?? name}-field-error-message`}
                    name={name}
                    testId={`${kebabCase(name)}-field-error-message`}
                />

                {footer}
            </>
        ),
        [label, labelContent, id, input, name, footer, useFloatingLabel],
    );

    return (
        <BasicFormField
            additionalClassName={additionalClassName}
            name={name}
            onChange={onChange}
            render={renderField}
        />
    );
};

export default memo(FormField);
