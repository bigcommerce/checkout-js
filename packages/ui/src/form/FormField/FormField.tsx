import { FieldProps } from 'formik';
import { kebabCase } from 'lodash';
import React, { FunctionComponent, memo, ReactNode, useCallback } from 'react';

import { BasicFormField } from '../BasicFormField';
import { FormFieldError } from '../FormFieldError';
import { Label } from '../Label';

export interface FormFieldProps {
    additionalClassName?: string;
    name: string;
    label?: ReactNode | ((fieldName: string) => ReactNode);
    labelContent?: ReactNode;
    footer?: ReactNode;
    id?: string;
    input(field: FieldProps<string>): ReactNode;
    onChange?(value: string): void;
}

const FormField: FunctionComponent<FormFieldProps> = ({
    additionalClassName,
    labelContent,
    label,
    onChange,
    footer,
    input,
    name,
    id,
}) => {
    const renderField = useCallback(
        (props: FieldProps<string>) => (
            <>
                {Boolean(label) && (typeof label === 'function' ? label(name) : label)}

                {Boolean(labelContent && !label) && (
                    <Label htmlFor={name} id={`${id ?? name}-label`}>
                        {labelContent}
                    </Label>
                )}

                {input(props)}

                <FormFieldError
                    errorId={`${id ?? name}-field-error-message`}
                    name={name}
                    testId={`${kebabCase(name)}-field-error-message`}
                />

                {footer}
            </>
        ),
        [label, labelContent, id, input, name, footer],
    );

    return (
        <BasicFormField
            additionalClassName={additionalClassName}
            name={name}
            onChange={onChange}
            render={renderField}
        />
    );
};

export default memo(FormField);
