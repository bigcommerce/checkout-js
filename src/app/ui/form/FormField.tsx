import { FieldProps } from 'formik';
import { kebabCase } from 'lodash';
import React, { memo, useCallback, Fragment, FunctionComponent, ReactNode } from 'react';

import BasicFormField from './BasicFormField';
import FormFieldError from './FormFieldError';
import Label from './Label';

export interface FormFieldProps {
    additionalClassName?: string;
    name: string;
    label?: ReactNode | ((fieldName: string) => ReactNode);
    labelContent?: ReactNode;
    footer?: ReactNode;
    input(field: FieldProps<string>): ReactNode;
    onChange?(value: string): void;
}

const FormField: FunctionComponent<FormFieldProps> = ({
    additionalClassName,
    labelContent,
    label,
    onChange,
    footer,
    input,
    name,
}) => {
    const renderField = useCallback(props => (
        <Fragment>
            { label && (typeof label === 'function' ? label(name) : label) }
            { labelContent && !label && <Label htmlFor={ name }>
                { labelContent }
            </Label> }

            { input(props) }

            <FormFieldError
                name={ name }
                testId={ `${kebabCase(name)}-field-error-message` }
            />

            { footer }
        </Fragment>
    ), [
        label,
        labelContent,
        input,
        name,
        footer,
    ]);

    return <BasicFormField
        additionalClassName={ additionalClassName }
        name={ name }
        onChange={ onChange }
        render={ renderField }
    />;
};

export default memo(FormField);
