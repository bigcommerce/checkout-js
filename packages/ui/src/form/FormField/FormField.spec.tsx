import { render } from 'enzyme';
import React from 'react';

import FormField from './FormField';

describe('FormField', () => {
    it('matches snapshot', () => {
        const props = {
            additionalClassName: 'test',
            labelContent: null,
            label: null,
            onChange: jest.fn(),
            footer: null,
            input: jest.fn(),
            name: 'test',
        };

        expect(render(<FormField {...props} />)).toMatchSnapshot();
    });

    it('matches snapshot when label is type of function', () => {
        const props = {
            additionalClassName: 'test',
            labelContent: null,
            label: jest.fn(),
            onChange: jest.fn(),
            footer: null,
            input: jest.fn(),
            name: 'test',
        };

        expect(render(<FormField {...props} />)).toMatchSnapshot();
    });

    it('matches snapshot when labelContent is provided', () => {
        const props = {
            additionalClassName: 'test',
            labelContent: jest.fn(),
            onChange: jest.fn(),
            footer: null,
            input: jest.fn(),
            name: 'test',
        };

        expect(render(<FormField {...props} />)).toMatchSnapshot();
    });
});
