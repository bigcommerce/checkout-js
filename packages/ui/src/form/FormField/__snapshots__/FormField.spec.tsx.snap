// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormField matches snapshot 1`] = `
<div
  class="form-field test"
/>
`;

exports[`FormField matches snapshot when label is type of function 1`] = `
<div
  class="form-field test"
/>
`;

exports[`FormField matches snapshot when labelContent is provided 1`] = `
<div
  class="form-field test"
>
  <label
    class="form-label optimizedCheckout-form-label"
    for="test"
    id="test-label"
  />
</div>
`;
